File: secondSeminar/src/main/java/org/sopt/secondSeminar/service/MemberService.java
package org.sopt.secondSeminar.service;

import jakarta.persistence.EntityNotFoundException;
import lombok.RequiredArgsConstructor;
import org.sopt.secondSeminar.Repository.MemberRepository;
import org.sopt.secondSeminar.domain.Member;
import org.sopt.secondSeminar.domain.SOPT;
import org.sopt.secondSeminar.dto.request.MemberCreateRequest;
import org.sopt.secondSeminar.dto.request.MemberProfileUpdateRequest;
import org.sopt.secondSeminar.dto.response.MemberGetResponse;
import org.sopt.secondSeminar.exception.ErrorMessage;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class MemberService {
    private final MemberRepository memberRepository;


    public MemberGetResponse getById(final Long memberId) {
        return MemberGetResponse.of(memberRepository.findById(memberId)
                .orElseThrow(() -> new EntityNotFoundException(ErrorMessage.MEMBER_NOT_FOUND_EXCEPTION.getMessage())));
    }


    public List<MemberGetResponse> getMembers() {
        return memberRepository.findAll()
                .stream()
                .map(MemberGetResponse::of)
                .collect(Collectors.toList());
    }

    @Transactional
    public String create(final MemberCreateRequest request) {
        Member member = memberRepository.save(Member.create(request));
        return member.getId().toString();
    }

    @Transactional
    public void updateSOPT(final Long memberId,
                           final MemberProfileUpdateRequest updateRequest) {
        Member member = memberRepository.findByIdOrThrow(memberId);
        member.updateSOPT(new SOPT(updateRequest.generation(), updateRequest.part()));
    }

    @Transactional
    public void delete(final Long memberId) {
        Member member = memberRepository.findByIdOrThrow(memberId);
        memberRepository.delete(member);
    }
}

File: secondKotlin/src/main/kotlin/org/sopt/secondKotlin/service/MemberService.kt
package org.sopt.secondKotlin.service;

import jakarta.persistence.EntityNotFoundException
import org.sopt.secondKotlin.domain.Member
import org.sopt.secondKotlin.dto.request.MemberCreateRequest
import org.sopt.secondKotlin.dto.request.MemberProfileUpdateRequest
import org.sopt.secondKotlin.dto.response.MemberGetResponse
import org.sopt.secondKotlin.repository.MemberRepository
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class MemberService(
    private val memberRepository: MemberRepository
) {

    @Transactional
    fun create(request: MemberCreateRequest): String {
        val member: Member = memberRepository.save(Member.create(request))
        return member.id.toString()
    }

    @Transactional
    fun updateSopt(
        memberId: Long,
        updateRequest: MemberProfileUpdateRequest
    ) {
        val member: Member =
            memberRepository.findByIdOrNull(memberId) ?: throw EntityNotFoundException("존재하지 않는 회원입니다.")
    }

    fun getById(
        memberId: Long
    ): MemberGetResponse {
        val member: Member =
            memberRepository.findByIdOrNull(memberId) ?: throw EntityNotFoundException("존재하지 않는 회원입니다.")
        return MemberGetResponse.of(member)
    }

    fun getMembers(): List<MemberGetResponse> {
        return memberRepository.findAll()
            .map { MemberGetResponse.of(it) }
            .toList() //kotlin의 it
    }

    @Transactional
    fun delete(
        memberId: Long
    ) {
        val member: Member =
            memberRepository.findByIdOrNull(memberId) ?: throw EntityNotFoundException("존재하지 않는 회원입니다.")
        memberRepository.delete(member)
    }
}
